ENIGMAsystem/SHELL/Graphics_Systems/Direct3D11/DX11vertex.cpp

```diff
@@ -16,23 +16,279 @@
 **/
 
 #include "Graphics_Systems/General/GSvertex_impl.h"
+#include "Graphics_Systems/General/GScolor_macros.h"
+
+#include "Bridges/General/DX11Context.h"
+
+#include <D3Dcompiler.h>
+
+namespace {
+const char* g_strVS =
+  "struct VertexInputType {\n"
+  "  float4 position : POSITION;\n"
+  "  float4 color : COLOR;\n"
+  "};\n"
+  "struct PixelInputType {\n"
+  "  float4 position : SV_POSITION;\n"
+  "  float4 color : COLOR;\n"
+  "};\n"
+  "PixelInputType VS(VertexInputType input) {\n"
+  "  PixelInputType output;\n"
+  "  output.position = input.position;\n"
+  "  output.color = input.color;\n"
+  "  return output;\n"
+  "}\n";
+
+const char* g_strPS =
+  "struct PixelInputType {\n"
+  "  float4 position : SV_POSITION;\n"
+  "  float4 color : COLOR;\n"
+  "};\n"
+  "float4 PS(PixelInputType input) : SV_TARGET {\n"
+  "  return input.color;\n"
+  "}\n";
+
+ID3D10Blob* pBlobVS = NULL;
+ID3D10Blob* pBlobPS = NULL;
+
+D3D_PRIMITIVE_TOPOLOGY primitive_types[] = {
+  static_cast<D3D_PRIMITIVE_TOPOLOGY>(0),
+  D3D11_PRIMITIVE_TOPOLOGY_POINTLIST,
+  D3D11_PRIMITIVE_TOPOLOGY_LINELIST,
+  D3D11_PRIMITIVE_TOPOLOGY_LINESTRIP,
+  D3D11_PRIMITIVE_TOPOLOGY_TRIANGLELIST,
+  D3D11_PRIMITIVE_TOPOLOGY_TRIANGLESTRIP,
+  static_cast<D3D_PRIMITIVE_TOPOLOGY>(0) // fans are deprecated
+};
+
+const char* semantic_names[] = { "POSITION", "COLOR", "NORMAL", "TEXCOORD", "BLENDWEIGHT",
+  "BLENDINDICES", "DEPTH", "TANGENT", "BINORMAL", "FOG", "SAMPLE" };
+
+DXGI_FORMAT dxgi_formats[] = {
+  DXGI_FORMAT_R32_FLOAT,
+  DXGI_FORMAT_R32G32_FLOAT,
+  DXGI_FORMAT_R32G32B32_FLOAT,
+  DXGI_FORMAT_R32G32B32A32_FLOAT,
+  DXGI_FORMAT_R8G8B8A8_UNORM,
+  DXGI_FORMAT_R8G8B8A8_UINT
+};
+size_t dxgi_format_sizes[] = {
+  sizeof(float) * 1,
+  sizeof(float) * 2,
+  sizeof(float) * 3,
+  sizeof(float) * 4,
+  sizeof(unsigned byte) * 4,
+  sizeof(unsigned byte) * 4
+};
+
+map<int, ID3D11Buffer*> vertexBufferPeers;
+map<int, ID3D11Buffer*> indexBufferPeers;
+}
 
 namespace enigma {
 
-void graphics_delete_vertex_buffer_peer(int buffer) {}
+void graphics_delete_vertex_buffer_peer(int buffer) {
+  vertexBufferPeers[buffer]->Release();
+  vertexBufferPeers.erase(buffer);
+}
+
+void graphics_delete_index_buffer_peer(int buffer) {
+  indexBufferPeers[buffer]->Release();
+  indexBufferPeers.erase(buffer);
+}
+
+void graphics_prepare_buffer(const int buffer, const bool isIndex) {
+  const bool dirty = isIndex ? indexBuffers[buffer]->dirty : vertexBuffers[buffer]->dirty;
```

Comment:
I'd just extract `const auto &buffer = isIndex ? indexBuffers[buffer] : vertexBuffers[buffer]`.

Reply:
That is not valid because those have different types, you are confusing indexBuffers/vertexBuffers with indexBufferPeers/vertexBufferPeers:
```
Graphics_Systems/Direct3D11/DX11vertex.cpp:103:78: error: conditional expression between distinct pointer types '__gnu_cxx::__alloc_traits<std::allocator<enigma::IndexBuffer*> >::value_type {aka enigma::IndexBuffer*}' and '__gnu_cxx::__alloc_traits<std::allocator<enigma::VertexBuffer*> >::value_type {aka enigma::VertexBuffer*}' lacks a cast
   const auto &buffer2 = isIndex ? indexBuffers[buffer] : vertexBuffers[buffer];
                                                                              ^
```

Reply:
However, after looking it over, I was able to extract `bufferPeers` now since they are the same type, I did not know it would be valid C++. See c075251ed90f8b1c3c667535343d257a555425d2

Reply:
You could also use a template, but don't worry about it.

Reply:
I finally got what you meant with this and didn't realize I could use templates in such a way that I can directly access members of the templated parameter if they are basically the same for each of the two types. I may be able to clean this up across all of the vertex implementations as well as some other cleanup, I'll save that for a future pull request.